readthedocs/projects/tasks/builds.py

```diff
@@ -0,0 +1,1020 @@
+        VersionLockedError,
+        ProjectBuildsSkippedError,
+        ConfigError,
+        YAMLParseError,
+    )
+
+    acks_late = True
+    track_started = True
+
+    # These values have to be dynamic based on project
+    time_limit = None
+    soft_time_limit = None
+
+    Request = BuildRequest
+
+    # TODO:
+    # 1. raise Reject(requeue=False) on duplicated builds
+    # 2. use a global `task_cls` (https://docs.celeryproject.org/en/latest/userguide/tasks.html#app-wide-usage) to logs actions
+    # 3. use CELERY_IMPORTS to register the tasks (https://docs.celeryproject.org/en/latest/userguide/configuration.html#std-setting-imports)
+    # 4. use CELERY_TASK_IGNORE_RESULT=True since we are not using the result at all
+
+    def _setup_sigterm(self):
+        def sigterm_received(*args, **kwargs):
+            log.warning('SIGTERM received. Waiting for build to stop gracefully after it finishes.')
+
+        # Do not send the SIGTERM signal to children (pip is automatically killed when
+        # receives SIGTERM and make the build to fail one command and stop build)
+        signal.signal(signal.SIGTERM, sigterm_received)
+
+    def _check_concurrency_limit(self):
+        try:
+            response = api_v2.build.concurrent.get(project__slug=self.project.slug)
+            concurrency_limit_reached = response.get('limit_reached', False)
+            max_concurrent_builds = response.get(
+                'max_concurrent',
+                settings.RTD_MAX_CONCURRENT_BUILDS,
+            )
+        except Exception:
+            log.exception(
+                'Error while hitting/parsing API for concurrent limit checks from builder.',
+                project_slug=self.project.slug,
+                version_slug=self.version.slug,
+            )
+            concurrency_limit_reached = False
+            max_concurrent_builds = settings.RTD_MAX_CONCURRENT_BUILDS
+
+        if concurrency_limit_reached:
+            # TODO: this could be handled in `on_retry` probably
+            log.warning(
+                'Delaying tasks due to concurrency limit.',
+                project_slug=self.project.slug,
+                version_slug=self.version.slug,
+            )
+
+            # This is done automatically on the environment context, but
+            # we are executing this code before creating one
+            api_v2.build(self.build['id']).patch({
+                'error': BuildMaxConcurrencyError.message.format(
+                    limit=max_concurrent_builds,
+                ),
+                'builder': socket.gethostname(),
+            })
+            self.retry(
+                exc=BuildMaxConcurrencyError,
+                throw=False,
+                # We want to retry this build more times
+                max_retries=25,
+            )
+
+    def _check_duplicated_build(self):
+        if self.build.get('status') == DuplicatedBuildError.status:
+            log.warning('NOOP: build is marked as duplicated.')
```

Comment:
Not sure this is correct, but "NOOP" might not be clear when we are reviewing logs outside the context of code.

```suggestion
            log.warning('Skipping build, build is marked as duplicated.')
```